### PR TITLE
fixup ruby warnings (v1.2.x)

### PR DESCRIPTION
**Pull request by eprothro**
This fixes various ruby warnings from uninitialized instance variables. These warnings cause noise for dependent gems and applications when running unit tests when printing warnings are enabled (a best practice for some).

### CODE REVIEW

File: lib/ione/future.rb [hunk 1]
```diff
@@ -327,7 +327,7 @@ def traverse(values, &block)
       # @param [Array<Ione::Future>] futures an array of futures whose values
       #   should be reduced
       # @param [Object] initial_value the initial value of the accumulator. When
-      #   nil (the default) the value of the first future will be used instead. 
+      #   nil (the default) the value of the first future will be used instead.
```
Comment (iconara) [inline lib/ione/future.rb]: Please don't make unrelated changes. It's not that I don't think that this space should be removed, it's just that it's just noise in the context of this PR. It will be removed when this line is changed the next time.

File: CHANGELOG.md [hunk 2]
```diff
@@ -0,0 +1,7 @@
+## 1.2.4
+
+Added:
+  - CHANGELOG.md
+
+Changed:
+  - Fixed various ruby warnings due to uninitialized instance variables
```
Comment (iconara) [inline CHANGELOG.md]: Please remove this file.

File: lib/ione/version.rb [hunk 3]
```diff
@@ -1,5 +1,5 @@
 # encoding: utf-8
 
 module Ione
-  VERSION = '1.2.3'.freeze
+  VERSION = '1.2.4'.freeze
```
Comment (iconara) [inline lib/ione/version.rb]: Don't force the version. The version number will be chosen after PRs have been merged. There might be other PRs that will be merged and released together that would require a minor version bump.

**Comment by iconara** [comment]
Thank you for making the changes I requested.

> To be clear, are you requesting individual commits for each warning that is fixed up?

No, but since you've already split everything up into one commit per file just leave it as it is.

What I had wanted to see from the beginning would be one commit that added the RSpec config that enabled warnings (that seems to have gotten lost now), and one commit with the fixes (and separate commits for the whitespace change, the version change and changelog, if those should have been included). Each commit should have had a commit message that explained _why_ the change was necessary or desirable. Your commit messages just say exactly what I can see in the diff. I need to understand why you want this in order to know whether or not it is the right way to do it, if it's a suitable change, etc., and to make it easier for me to pick out the pieces that I think should be merged (I could just have skipped merging the commits with the version bump and the whitespace change, for example).

**Comment by iconara** [comment]
This prints a wall of messages similar to

```
ione/spec/integration/ssl_spec.rb:83: warning: possibly useless use of == in void context
```

when run with warnings enabled. Every use of `….should == …` needs to be rewritten to `….should eq(…)`.

If you really think this is a big deal, you're welcome to add that to this PR. I don't see the point in fixing one problem and not the other.

**Comment by eprothro** [comment]
@iconara ok, thanks for explaining. Would you like me to squash the commits and give a more explanatory commit message?

I reverted the rspec config change because I didn't intend to leave it in your project. I like to run my suite with warnings on, doesn't mean you need to :).

**Comment by eprothro** [comment]
Regarding the rspec related warnings: I don't feel the same way about those as they won't affect clients that use your gem (since they only come when running `ione`'s suite.

Only 3 of the variables I fixed up were actually printing warnings in my test suite, but I fixed up the ones I saw printed when running the `ione` suite as well.

**Comment by iconara** [comment]
I don't mind enabling warnings in this project, as long as that doesn't mean having to look at a lot of warnings. I'll merge this if you go the whole way and make this project run without warnings. I'm not merging it if all you want to do is fix warnings in your project.

**Comment by eprothro** [comment]
I'm happy to fix warnings for this project when I have time, I don't at the moment.

This PR does more than just fix warnings for my project, if fixes all the warnings that I'm aware of that would affect any project using this gem.

**Comment by eprothro** [comment]
@iconara any thoughts on my last comment? I believe this PR represents a positive change for many projects, not just my own. I propose we open a separate issue for resolving all internal warnings for this project, and we can get to it as time allows.

**Comment by iconara** [comment]
I think it's a great idea to fix the code so that there are no warnings. I don't see the need for it myself (hence the current state of things), but if you feel strongly for it I'm definitely not against the change. However, just fixing the current warnings is not enough. It might fix your immediate problem, but the next PR might add code that triggers new warnings, and we'd be back here again. There needs to be a mechanism that makes it easy to avoid that. Enabling warnings in Ione's own test suite seems like the best way – but that requires the test suite to be fixed to not trigger warnings.

Have you considered what happens when v1.3.0 is released (it's been a long time coming, but it will be released someday)? I would apply this PR to that branch as well, but it's changed enough that it might trigger other warnings. Wouldn't you expect me to ensure that that release wouldn't trigger new warnings?

I'm saying this so that you understand what it is that you're asking: you want me to keep Ione warning free now, and in future releases. I can support that, but not if I have to do the work to make it possible to run the tests with warnings on.

Help me get to a warning free test suite and I'll commit to keeping future releases warning free.

**Comment by eprothro** [comment]
I agree, thanks for taking the time to say more.

I'll try to make time for this next week.

**Comment by eprothro** [comment]
Hi Theo, warnings should be eliminated for this branch (1.2.x) now.

I introduced static DH params from a pregenerated DH param PEM so the specs wouldn't take the time (10+sec) to generate the DH on the fly.

I'll try to apply to a branch from master tomorrow or next week and generate a separate PR unless you would like that done a different way.

**Comment by iconara** [comment]
Looks pretty good (I would have preferred `should eq(…)` over `should eq …`, but I can live with it it). There's still some SSL warnings in JRuby (and the JRuby tests deadlock in Travis, but I don't think that's your fault, not sure what's up with that).

**Comment by eprothro** [comment]
@iconara ah, sorry, I misread that to say 'don't worry about the jRuby warnings'. Will look at them.

**Comment by iconara** [comment]
Now the JRuby tests error out completely. They aren't even runnable.

Is removing `Gemfile.lock` necessary for this PR? Otherwise could you please stop applying your ideas of how projects should be structured. I'm not saying it's a good idea to include it, or that it must be included (things were different back when it was added), but this is not the place to make those kinds of changes, and it is definitely not your decision. Stick to what this PR is about.

**Comment by iconara** [comment]
Can you please also write a proper commit message for 1a28db344fa597bf45ca15314ecc80c22fd526e8? It does things with certificates and SSL config but doesn't explain why, what or anything that would make it possible to understand why that change was necessary. I've tried googling the `:SSLTmpDhCallback` parameter but there's no documentation. I would really like to know what's going on, but you've provided no explanation and no context.

**Comment by iconara** [comment]
Closing in favour of #48

**Comment by eprothro** [comment]
@iconara apologies that I wasn't able to do this to your liking. It was never my intent to argue or make decisions for you.

I removed the gemfile.lock to simplify my own development of switching ruby version and bundling, etc, and included that change in a separate commit, so you could ignore if you didn't like it, per your previous instructions. This is also why I pushed it up and asked if you were for that change before I continue to fix the side effects it caused (no point if you didn't like that decision).

The `temp_dh_callback` is a way for the server to generate DH cipher params when they are required for TLS.  `SSLTmpDhCallback` is the WEBrick option.

http://ruby-doc.org/stdlib-1.9.3/libdoc/openssl/rdoc/OpenSSL/SSL/SSLContext.html#tmp_dh_callback-attribute-method

https://ruby-doc.org/stdlib-2.1.3/libdoc/webrick/rdoc/WEBrick/Config.html

http://security.stackexchange.com/questions/94390/whats-the-purpose-of-dh-parameters
